dockerfiles/init/modules/codenvy/templates/rsyncrestore.sh.erb

```diff
@@ -10,33 +8,76 @@ DST_USR_ID=${5}
 DST_GRP_ID=${6}
 SRC_USER_NAME=${7}
 
-# ensure existing trailing slash to overwrite permissions of destination folder too
-if [[ ${SRC_FOLDER} != */ ]]; then
-   SRC_FOLDER=${SRC_FOLDER}"/"
-fi
-if [[ ${DST_FOLDER} != */ ]]; then
-   DST_FOLDER=${DST_FOLDER}"/"
-fi
-
-# ensure ssh key has correct permissions
+# Key for SSH connection to sync files
+SSH_KEY=/opt/codenvy-data/conf/ssh/key.pem
+# When Codenvy runs on windows we have issue since key doesn't have permissions 0600.
+# With another permissions on key SSH doesn't work.
+# So we copy key into another place that should be inside of container - so it is Linux and permissions are supported.
+# And apply needed permission to copied file. Copying is performed lazily, so it is skipped when it was done previously.
 if [[ ! -f /opt/rsync_key ]]; then
     cp $SSH_KEY /opt/rsync_key
 fi
 if [[ $(stat -c %a /opt/rsync_key) != 600 ]]; then
     chmod 0600 /opt/rsync_key
 fi
 
+##### Set by puppet #####
 RSYNC_RESTORE_BWLIMIT=<%= scope.lookupvar('codenvy::rsync_restore_bwlimit') %>
 SSH_LOG_LEVEL=<%= scope.lookupvar('codenvy::rsync_ssh_log_level') %>
-SSH_OPTIONS="-i /opt/rsync_key -l ${SRC_USER_NAME} -o StrictHostKeyChecking=no -o PasswordAuthentication=no -o LogLevel=${SSH_LOG_LEVEL} -p ${DST_PORT}"
+#########################
 
+##### SSH options #####
+SSH_OPTIONS=""
+# Add SSH connection options
+SSH_OPTIONS+=" -i /opt/rsync_key -l ${SRC_USER_NAME} -p ${DST_PORT}"
+# Disable password authentication since we use key-based auth
+SSH_OPTIONS+=" -o PasswordAuthentication=no"
+# Disable hosts fingerprint checking because it may fail due to
+# starting different containers on the same ports after some period
+SSH_OPTIONS+=" -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"
+# Set SSH logging level to make it possible to investigate problems
+SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
+#######################
+
+# We need root permissions in target container to change ownership of files.
+# If user in that container has access to sudo we use it.
+# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
 RSYNC_COMMAND=$(ssh ${SSH_OPTIONS} ${DST_HOST} \
             "if hash sudo 2>/dev/null;then echo 'sudo rsync';else echo 'rsync';fi")
 
-rsync --quiet --delete --partial --links --safe-links --owner --group --recursive --times --human-readable --bwlimit=${RSYNC_RESTORE_BWLIMIT} --executability \
-      --include='.codenvy/' \
-      --filter=':- .gitignore' \
+##### Rsync options #####
+RSYNC_OPTIONS=""
+RSYNC_OPTIONS+=" --quiet"
+RSYNC_OPTIONS+=" --recursive"
+# Throughput limit for rsync
+RSYNC_OPTIONS+=" --bwlimit=${RSYNC_RESTORE_BWLIMIT}"
+# Sync modification timestamps to optimise transfer of not modified files.
+RSYNC_OPTIONS+=" --times"
+# Do not remove partially transferred files if transfer is interrupted. Make subsequent transfers faster.
+RSYNC_OPTIONS+=" --partial"
```

Comment:
I think we would fail and remove container if first transfer failed, so it useless to add this switch